atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultBucketAssigner.java

```diff
@@ -0,0 +1,235 @@
+            } else {
+                return IterationState.of(State.INCOMPLETE, bucketIdentifier, waitingClose);
+            }
+        }
+
+        @Override
+        public IterationState visit(BucketAssignerState.ClosingFromOpen closingFromOpen) {
+            metrics.bucketAssignerState(closingFromOpen);
+            return tryTransitionToStartAfterClosingBuckets(
+                    Optional.of(TimestampRange.openBucket(closingFromOpen.startTimestampInclusive())),
+                    TimestampRange.of(
+                            closingFromOpen.startTimestampInclusive(), closingFromOpen.endTimestampExclusive()),
+                    closingFromOpen);
+        }
+
+        @Override
+        public IterationState visit(BucketAssignerState.ImmediatelyClosing immediateClose) {
+            metrics.bucketAssignerState(immediateClose);
+            return tryTransitionToStartAfterClosingBuckets(
+                    Optional.empty(),
+                    TimestampRange.of(immediateClose.startTimestampInclusive(), immediateClose.endTimestampExclusive()),
+                    immediateClose);
+        }
+
+        private IterationState tryTransitionToStartAfterClosingBuckets(
+                Optional<TimestampRange> oldTimestampRange,
+                TimestampRange newTimestampRange,
+                BucketAssignerState lastState) {
+            WriteState writeState =
+                    bucketWriter.writeToAllBuckets(bucketIdentifier, oldTimestampRange, newTimestampRange);
+
+            if (writeState == WriteState.FAILED_CAS) {
+                return IterationState.of(State.CONTENTION_ON_WRITES, bucketIdentifier, lastState);
+            }
+
+            BucketAssignerState start = BucketAssignerState.start(newTimestampRange.endExclusive());
+            sweepBucketAssignerStateMachineTable.updateStateMachineForBucketAssigner(
+                    BucketStateAndIdentifier.of(bucketIdentifier, lastState),
+                    BucketStateAndIdentifier.of(bucketIdentifier + 1, start));
+            return IterationState.of(State.CLOSED, bucketIdentifier, start);
+        }
+
+        private IterationState tryTransitionAndExecuteNextState(
+                BucketAssignerState currentState, BucketAssignerState newState) {
+            sweepBucketAssignerStateMachineTable.updateStateMachineForBucketAssigner(
+                    BucketStateAndIdentifier.of(bucketIdentifier, currentState),
+                    BucketStateAndIdentifier.of(bucketIdentifier, newState));
+            return newState.accept(this);
+        }
+    }
+
+    @Value.Immutable
+    interface IterationState {
+        @Value.Parameter
+        State iterationState();
+
+        @Value.Parameter
+        long bucketIdentifier();
+
+        @Value.Parameter
+        BucketAssignerState bucketAssignerState();
+
+        // TODO(mdaudali): Re-evaluate if this is a useful metric or not.
+        enum State {
```

Comment:
let's revisit the name on this one - it reads too much like `BucketAssignerState` which is the actual state machine state

Reply:
(handled by `IterationResult` / `OperationResult` change)